### PR TITLE
filter bundles-containing by ACL

### PR DESCRIPTION
**Pull request by mhilton**
Add a filter on bundles containing so that bundles aren't listed if the requester cannot access them.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1106/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1107/
Test PASSed.

File: internal/v4/relations_test.go [hunk 1]
```diff
@@ -766,6 +768,53 @@ func (s *RelationsSuite) TestMetaBundlesContaining(c *gc.C) {
...
+		// The charm-bundle relations are retrieved from the entities
+		// collection, without accessing the blob store.
+		err := s.store.AddBundle(b, charmstore.AddParams{
+			URL:      url,
+			BlobName: "blobName",
+			BlobHash: fakeBlobHash,
+			BlobSize: fakeBlobSize,
+		})
+		c.Assert(err, gc.IsNil)
+		if url.URL.Name == "useless" {
```
Comment (frankban) [inline internal/v4/relations_test.go:785]: Add a comment about removing read permission to everyone for the useless bundle?

File: internal/v4/relations_test.go [hunk 2]
```diff
@@ -766,6 +768,53 @@ func (s *RelationsSuite) TestMetaBundlesContaining(c *gc.C) {
...
+		URL:      rurl,
+		BlobName: "blobName",
+		BlobHash: fakeBlobHash,
+		BlobSize: fakeBlobSize,
+	})
+	c.Assert(err, gc.IsNil)
+	err = s.store.SetPerms(&rurl.URL, "read", params.Everyone, rurl.URL.User)
+	c.Assert(err, gc.IsNil)
+
+	// Perform the request and ensure the response is what we expect.
```
Comment (frankban) [inline internal/v4/relations_test.go]: Perhaps suggest that we expect the useless bundle to not be included in the results.

**Comment by frankban** [comment]
:+1: with only two comment requests.

File: internal/v4/relations.go [hunk 3]
```diff
@@ -264,6 +264,16 @@ func (h *ReqHandler) metaBundlesContaining(entity *mongodoc.Entity, id *router.R
...
 	// TODO(rog) make this concurrent.
 	for _, e := range entities {
+		rurl := router.ResolvedURL{
+			URL:                 *e.URL,
+			PromulgatedRevision: -1,
+		}
+		if e.PromulgatedURL != nil {
+			rurl.PromulgatedRevision = e.PromulgatedURL.Revision
+		}
+		if err := h.AuthorizeEntity(&rurl, req); err != nil {
```
Comment (rogpeppe) [inline internal/v4/relations.go]: s/&rurl/charmstore.EntityResolvedURL(e)/

and delete the rest.

Comment (rogpeppe) [inline internal/v4/relations.go]: Perhaps add a comment:
// Ignore entities for which the user does not have permission.

?

**Comment by rogpeppe** [comment]
LGTM modulo a couple of suggestions.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1108/
Test PASSed.

**Comment by mhilton** [comment]
:shipit: thanks for the reviews

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**